### PR TITLE
[flutter_tools] Apply --no-causal-async-stacks and --lazy-async-stacks to profile/release builds

### LINKED ISSUE #48725: Enable lazy async stack traces via tool feature
The additional flags `--no-causal-async-stacks` and  `--lazy-async-stacks` can drastically improve the performance of async/await operations, at the cost of slightly different stack traces.

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes https://github.com/flutter/flutter/issues/48725

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Based on my current understanding of the breaking change policy, this is not a breaking change.

File: packages/flutter_tools/lib/src/base/build.dart [hunk 1]
```diff
@@ -156,6 +155,13 @@ class AOTSnapshotter {
       genSnapshotArgs.add('--no-use-integer-division');
     }
 
+    // Optimization arguments.
+    genSnapshotArgs.addAll(<String>[
+      // Faster async/await
+      '--no-causal-async-stacks',
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/build.dart]: We should also update the flags to gen_snapshot in `fuchsia/fuchsia_build.dart` and `dart_component.gni` in the Fuchsia tree to ensure we're doing the same thing everywhere. @iskakaushik

File: packages/flutter_tools/lib/src/compile.dart [hunk 2]
```diff
@@ -650,7 +646,7 @@ class DefaultResidentCompiler implements ResidentCompiler {
       sdkRoot,
       '--incremental',
       '--target=$targetModel',
-      '-Ddart.developer.causal_async_stacks=$causalAsyncStacks',
+      '-Ddart.developer.causal_async_stacks=${buildMode == BuildMode.debug}',
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/compile.dart]: This leads me to notice that we need to unify the flags used here with the flags used in `fuchsia_kernel_compiler.dart` either by keeping them in sync manually or by eliminating `fuchsia_kernel_compiler.dart` by merging its functionality into this file.

**Review by zanderso — COMMENTED** [review]
Did you spot-check some device lab tests?

**Comment by jonahwilliams** [comment]
@zanderso no I haven't checked any. I'm not aware of any stack trace parsing tests

File: packages/flutter_tools/lib/src/compile.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/compile.dart]: It looks like the kernel_compiler and frontend_server support a slightly different set of flags and have slightly different output behavior. Is there any reason we can't standardize on the frontend_server instead?

Comment (zanderso) [reply packages/flutter_tools/lib/src/compile.dart]: Ahh right. Functionality of the Fuchsia kernel compiler would need to be integrated into the frontend_server, so there's probably not much we can do right now other than manually try to keep the flags in-sync.

**Comment by iskakaushik** [comment]
There is a stacktrace parsing test in topaz that i'm aware of, i've notified the owner to watch out for changes: fxb/44622, we might need to do a manual roll.

**Comment by jonahwilliams** [comment]
This would only effect "release/profile" mode stack traces, so depending on how that test is running it might not be bothered

**Comment by jonahwilliams** [comment]
Looks like gen_snapshot invocation for fuchsia already had --no-causal-async-stacks applied

File: packages/flutter_tools/lib/src/fuchsia/fuchsia_build.dart [hunk 3]
```diff
@@ -98,7 +98,8 @@ Future<void> _genSnapshot(
 
   final List<String> command = <String>[
     genSnapshot,
-    '--no_causal_async_stacks',
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/fuchsia/fuchsia_build.dart]: I'm not sure if this matters...

Comment (zanderso) [reply packages/flutter_tools/lib/src/fuchsia/fuchsia_build.dart]: If dashes work for other flags, they should also work for these flags.

**Review by zanderso — APPROVED** [review]

**Pull request merged**